File: api/routes/playlist.js
const express = require('express');
const router = express.Router();
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { v4: uuidv4 } = require('uuid');
const archiver = require('archiver');

const { TEMP_DIRS, SAFETY_LIMITS, PLAYLIST_DOWNLOAD_EXPIRY } = require('../config/constants');
const {
  activeDownloads,
  activeProcesses,
  activeJobsByType,
  asyncJobs,
  botDownloads,
  registerClient,
  linkJobToClient,
  unlinkJobFromClient,
  getClientJobCount,
  sendProgress
} = require('../services/state');

const { downloadViaCobalt } = require('../services/cobalt');
const { downloadViaYtdlp, getPlaylistInfo } = require('../services/downloader');
const { processVideo } = require('../services/processor');
const { validateUrl } = require('../utils/validation');
const { toUserError } = require('../utils/errors');
const { cleanupJobFiles, sanitizeFilename } = require('../utils/files');
const discordAlerts = require('../discord-alerts');

router.post('/api/playlist/start', express.json(), async (req, res) => {
  const {
    url, format = 'video', quality = '1080p', container = 'mp4',
    audioFormat = 'mp3', audioBitrate = '320', clientId
  } = req.body;

  const urlCheck = validateUrl(url);
  if (!urlCheck.valid) return res.status(400).json({ error: urlCheck.error });

  if (clientId) {
    const clientJobs = getClientJobCount(clientId);
    if (clientJobs >= SAFETY_LIMITS.maxJobsPerClient) {
      return res.status(429).json({ error: `Too many active jobs. Maximum ${SAFETY_LIMITS.maxJobsPerClient} concurrent jobs per user.` });
    }
  }

  const jobId = uuidv4();
  const isAudio = format === 'audio';
  const outputExt = isAudio ? audioFormat : container;

  if (clientId) {
    registerClient(clientId);
    linkJobToClient(jobId, clientId);
  }

  const job = {
    status: 'starting',
    progress: 0,
    message: 'getting playlist info...',
    createdAt: Date.now(),
    url,
    format: outputExt,
    type: 'playlist',
    playlistTitle: null,
    totalVideos: 0,
    videosCompleted: 0,
    currentVideo: 0,
    currentVideoTitle: '',
    failedVideos: [],
    failedCount: 0,
    downloadToken: null,
    fileName: null,
    fileSize: null
  };

  asyncJobs.set(jobId, job);
  res.json({ jobId });

  processPlaylistAsync(jobId, job, url, isAudio, audioFormat, outputExt, quality, container, audioBitrate);
});

async function processPlaylistAsync(jobId, job, url, isAudio, audioFormat, outputExt, quality, container, audioBitrate) {
  const playlistDir = path.join(TEMP_DIRS.playlist, jobId);
  if (!fs.existsSync(playlistDir)) fs.mkdirSync(playlistDir, { recursive: true });

  const processInfo = { cancelled: false, process: null, tempDir: playlistDir };
  activeProcesses.set(jobId, processInfo);
  activeJobsByType.playlist++;
  console.log(`[Queue] Async playlist started. Active: ${JSON.stringify(activeJobsByType)}`);

  try {
    const playlistInfo = await getPlaylistInfo(url);

    if (playlistInfo.count > SAFETY_LIMITS.maxPlaylistVideos) {
      throw new Error(`Playlist too large. Maximum ${SAFETY_LIMITS.maxPlaylistVideos} videos allowed. This playlist has ${playlistInfo.count} videos.`);
    }

    const totalVideos = playlistInfo.count;
    const playlistTitle = playlistInfo.title;

    job.status = 'downloading';
    job.playlistTitle = playlistTitle;
    job.totalVideos = totalVideos;
    job.message = `found ${totalVideos} videos`;

    sendProgress(jobId, 'playlist-info', `Found ${totalVideos} videos in playlist`, 0, {
      playlistTitle, totalVideos, currentVideo: 0, currentVideoTitle: '',
      format: isAudio ? audioFormat : `${quality} ${container}`
    });

    const downloadedFiles = [];
    const failedVideos = [];

    for (let i = 0; i < playlistInfo.entries.length; i++) {
      if (processInfo.cancelled) throw new Error('Download cancelled');
      if (processInfo.finishEarly) {
        console.log(`[${jobId}] Finishing early after ${downloadedFiles.length} videos`);
        break;
      }

      const entry = playlistInfo.entries[i];
      const videoNum = i + 1;
      const videoTitle = entry.title || `Video ${videoNum}`;
      const videoUrl = entry.url || (entry.id ? `https://www.youtube.com/watch?v=${entry.id}` : null);
      if (!videoUrl && !entry.id) continue;

      const safeTitle = sanitizeFilename(videoTitle).substring(0, 100);
      const videoFile = path.join(playlistDir, `${String(videoNum).padStart(3, '0')} - ${safeTitle}.${outputExt}`);

      job.currentVideo = videoNum;
      job.currentVideoTitle = videoTitle;
      job.progress = ((videoNum - 1) / totalVideos) * 100;
      job.message = `downloading ${videoNum}/${totalVideos}: ${videoTitle}`;

      sendProgress(jobId, 'downloading', `Downloading ${videoNum}/${totalVideos}: ${videoTitle}`,
        job.progress, {
          playlistTitle, totalVideos, currentVideo: videoNum, currentVideoTitle: videoTitle,
          format: isAudio ? audioFormat : `${quality} ${container}`,
          failedVideos: failedVideos.slice(-50), failedCount: failedVideos.length
        });

      try {
        const actualVideoUrl = videoUrl || `https://www.youtube.com/watch?v=${entry.id}`;
        const isYouTubeVideo = actualVideoUrl.includes('youtube.com') || actualVideoUrl.includes('youtu.be');
        let tempPath = null;

        if (isYouTubeVideo) {
          const videoJobId = `${jobId}-v${videoNum}`;
          const abortController = new AbortController();
          processInfo.abortController = abortController;

          try {
            const cobaltResult = await downloadViaCobalt(actualVideoUrl, videoJobId, isAudio,
              (progress) => {
                const overallProgress = ((videoNum - 1) / totalVideos * 100) + (progress / totalVideos);
                job.progress = overallProgress;
                sendProgress(jobId, 'downloading', `Downloading ${videoNum}/${totalVideos}: ${videoTitle} (${progress}%)`,
                  overallProgress, { playlistTitle, totalVideos, currentVideo: videoNum, currentVideoTitle: videoTitle, videoProgress: progress,
                    format: isAudio ? audioFormat : `${quality} ${container}`, failedVideos: failedVideos.slice(-50), failedCount: failedVideos.length });
              }, abortController.signal, { outputDir: playlistDir, maxRetries: 3, retryDelay: 2000 });
            tempPath = cobaltResult.filePath;
          } catch (cobaltErr) {
            if (cobaltErr.message === 'Cancelled') throw cobaltErr;
            failedVideos.push({ num: videoNum, title: videoTitle, reason: toUserError(cobaltErr.message) });
            job.failedVideos = failedVideos.slice(-50);
            job.failedCount = failedVideos.length;
            continue;
          }
        }

        if (!tempPath && !isYouTubeVideo) {
          const result = await downloadViaYtdlp(actualVideoUrl, `temp_${videoNum}`, {
            isAudio, quality, container,
            tempDir: playlistDir,
            processInfo,
            onProgress: (progress, speed, eta) => {
              const overallProgress = ((videoNum - 1) / totalVideos * 100) + (progress / totalVideos);
              job.progress = overallProgress;
              if (speed) job.speed = speed;
              if (eta) job.eta = eta;
              sendProgress(jobId, 'downloading', `Downloading ${videoNum}/${totalVideos}: ${videoTitle} (${progress.toFixed(0)}%)`,
                overallProgress, { playlistTitle, totalVideos, currentVideo: videoNum, currentVideoTitle: videoTitle, videoProgress: progress, speed, eta });
            }
          });
          tempPath = result.path;
        }

        if (tempPath && fs.existsSync(tempPath)) {
          sendProgress(jobId, 'processing', `Processing ${videoNum}/${totalVideos}: ${videoTitle}`,
            ((videoNum - 0.5) / totalVideos) * 100, { playlistTitle, totalVideos, currentVideo: videoNum, currentVideoTitle: videoTitle });

          const processed = await processVideo(tempPath, videoFile, {
            isAudio, audioFormat, audioBitrate, container, jobId
          });

          if (processed.skipped && tempPath !== videoFile) {
            fs.renameSync(tempPath, videoFile);
          }

          downloadedFiles.push(videoFile);
          job.videosCompleted = downloadedFiles.length;
          console.log(`[${jobId}] Video ${videoNum} complete`);
        }
      } catch (err) {
        console.error(`[${jobId}] Error downloading video ${videoNum}:`, err.message);
        failedVideos.push({ num: videoNum, title: videoTitle, reason: toUserError(err.message) });
        job.failedVideos = failedVideos.slice(-50);
        job.failedCount = failedVideos.length;
        if (err.message === 'Cancelled' || err.message === 'Download cancelled') throw err;
      }
    }

    if (downloadedFiles.length === 0) throw new Error('No videos were successfully downloaded');

    job.status = 'zipping';
    job.progress = 95;
    job.message = `creating zip with ${downloadedFiles.length} videos...`;

    sendProgress(jobId, 'zipping', `Creating zip file with ${downloadedFiles.length} videos...`, 95, {
      playlistTitle, totalVideos, downloadedCount: downloadedFiles.length
    });

    const zipPath = path.join(TEMP_DIRS.playlist, `${jobId}.zip`);
    const safePlaylistName = sanitizeFilename(playlistTitle || 'playlist');

    await new Promise((resolve, reject) => {
      const output = fs.createWriteStream(zipPath);
      const archive = archiver('zip', { zlib: { level: 5 } });
      output.on('close', resolve);
      archive.on('error', reject);
      archive.pipe(output);
      downloadedFiles.forEach(filePath => archive.file(filePath, { name: path.basename(filePath) }));
      archive.finalize();
    });

    try {
      if (fs.existsSync(playlistDir)) fs.rmSync(playlistDir, { recursive: true, force: true });
    } catch {}

    const stat = fs.statSync(zipPath);
    const downloadToken = crypto.randomBytes(32).toString('hex');
    const fileName = `${safePlaylistName}.zip`;

    botDownloads.set(downloadToken, {
      filePath: zipPath, fileName, fileSize: stat.size,
      mimeType: 'application/zip', createdAt: Date.now(),
      downloaded: false, isWebPlaylist: true
    });

    job.status = 'complete';
    job.progress = 100;
    job.message = `${downloadedFiles.length} videos ready to download`;
    job.downloadToken = downloadToken;
    job.fileName = fileName;
    job.fileSize = stat.size;
    job.failedVideos = failedVideos.slice(-50);
    job.failedCount = failedVideos.length;

    sendProgress(jobId, 'complete', `${downloadedFiles.length} videos ready!`, 100, {
      playlistTitle, totalVideos, downloadedCount: downloadedFiles.length,
      failedVideos: failedVideos.slice(-50), failedCount: failedVideos.length,
      downloadToken
    });

    activeProcesses.delete(jobId);
    activeJobsByType.playlist--;
    unlinkJobFromClient(jobId);
    console.log(`[Queue] Async playlist complete. Active: ${JSON.stringify(activeJobsByType)}`);

  } catch (err) {
    console.error(`[${jobId}] Async playlist error:`, err.message);
    discordAlerts.downloadFailed('Playlist Download Error', 'Playlist download failed.', { jobId, error: err.message });

    job.status = 'error';
    job.message = toUserError(err.message || 'Playlist download failed');

    if (!processInfo.cancelled) {
      sendProgress(jobId, 'error', toUserError(err.message || 'Playlist download failed'));
    }

    activeProcesses.delete(jobId);
    activeJobsByType.playlist--;
    unlinkJobFromClient(jobId);
    console.log(`[Queue] Async playlist error. Active: ${JSON.stringify(activeJobsByType)}`);

    try {
      if (fs.existsSync(playlistDir)) fs.rmSync(playlistDir, { recursive: true, force: true });
    } catch {}
  }
}

router.get('/api/playlist/status/:jobId', (req, res) => {
  const { jobId } = req.params;
  const job = asyncJobs.get(jobId);
  if (!job) return res.status(404).json({ error: 'Job not found' });

  res.json({
    status: job.status,
    progress: job.progress,
    message: job.message,
    playlistTitle: job.playlistTitle,
    totalVideos: job.totalVideos,
    videosCompleted: job.videosCompleted,
    currentVideo: job.currentVideo,
    currentVideoTitle: job.currentVideoTitle,
    failedVideos: job.failedVideos,
    failedCount: job.failedCount,
    downloadToken: job.downloadToken,
    fileName: job.fileName,
    fileSize: job.fileSize,
    speed: job.speed,
    eta: job.eta
  });
});

router.get('/api/playlist/download/:token', (req, res) => {
  const { token } = req.params;
  const data = botDownloads.get(token);

  if (!data) return res.status(404).json({ error: 'Download not found or expired' });
  if (!fs.existsSync(data.filePath)) {
    botDownloads.delete(token);
    return res.status(404).json({ error: 'File no longer available' });
  }

  const stat = fs.statSync(data.filePath);
  const asciiFilename = data.fileName.replace(/[^\x20-\x7E]/g, '_');

  res.setHeader('Content-Type', data.mimeType);
  res.setHeader('Content-Length', stat.size);
  res.setHeader('Content-Disposition', `attachment; filename="${asciiFilename}"; filename*=UTF-8''${encodeURIComponent(data.fileName)}`);

  const stream = fs.createReadStream(data.filePath);
  stream.pipe(res);

  stream.on('close', () => {
    data.downloaded = true;
    setTimeout(() => {
      if (botDownloads.has(token)) {
        fs.unlink(data.filePath, () => {});
        botDownloads.delete(token);
        console.log(`[Playlist] Token ${token.slice(0, 8)}... cleaned up after download`);
      }
    }, 30000);
  });
});

setInterval(() => {
  const now = Date.now();
  for (const [token, data] of botDownloads.entries()) {
    if (data.isWebPlaylist && now - data.createdAt > PLAYLIST_DOWNLOAD_EXPIRY) {
      console.log(`[Playlist] Download token ${token.slice(0, 8)}... expired`);
      if (data.filePath && fs.existsSync(data.filePath)) {
        fs.unlink(data.filePath, () => {});
      }
      botDownloads.delete(token);
    }
  }
}, 60000);

module.exports = router;

File: internal/routes/playlist.go
package routes

import (
	"archive/zip"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/coah80/yoink/internal/alerts"
	"github.com/coah80/yoink/internal/config"
	"github.com/coah80/yoink/internal/services"
	"github.com/coah80/yoink/internal/util"
)

func PlaylistRoutes(r chi.Router) {
	r.Post("/api/playlist/start", handlePlaylistStart)
	r.Get("/api/playlist/status/{jobId}", handlePlaylistStatus)
	r.Get("/api/playlist/download/{token}", handlePlaylistDownload)
}

func handlePlaylistStart(w http.ResponseWriter, r *http.Request) {
	var body struct {
		URL          string `json:"url"`
		Format       string `json:"format"`
		Quality      string `json:"quality"`
		Container    string `json:"container"`
		AudioFormat  string `json:"audioFormat"`
		AudioBitrate string `json:"audioBitrate"`
		ClientID     string `json:"clientId"`
		ResumeFrom   int    `json:"resumeFrom"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		respondJSON(w, 400, map[string]string{"error": "Invalid request body"})
		return
	}

	if body.Format == "" {
		body.Format = "video"
	}
	if body.Quality == "" {
		body.Quality = "1080p"
	}
	if body.Container == "" {
		body.Container = "mp4"
	}
	if body.AudioFormat == "" {
		body.AudioFormat = "mp3"
	}
	if body.AudioBitrate == "" {
		body.AudioBitrate = "320"
	}
	if body.ResumeFrom < 1 {
		body.ResumeFrom = 1
	}

	check := util.ValidateURL(body.URL)
	if !check.Valid {
		respondJSON(w, 400, map[string]string{"error": check.Error})
		return
	}

	if body.ClientID != "" {
		if services.Global.GetClientJobCount(body.ClientID) >= config.MaxJobsPerClient {
			respondJSON(w, 429, map[string]string{"error": fmt.Sprintf("Too many active jobs. Maximum %d concurrent jobs per user.", config.MaxJobsPerClient)})
			return
		}
	}

	jobCheck := services.Global.CanStartJob("playlist")
	if !jobCheck.OK {
		respondJSON(w, 503, map[string]string{"error": jobCheck.Reason})
		return
	}

	jobID := uuid.New().String()
	isAudio := body.Format == "audio"
	outputExt := body.Container
	if isAudio {
		outputExt = body.AudioFormat
	}

	if body.ClientID != "" {
		services.Global.RegisterClient(body.ClientID)
		services.Global.LinkJobToClient(jobID, body.ClientID)
	}

	job := &services.AsyncJob{
		Status:    "starting",
		Message:   "getting playlist info...",
		CreatedAt: time.Now(),
		URL:       body.URL,
		Format:    outputExt,
		Type:      "playlist",
	}
	services.Global.SetAsyncJob(jobID, job)

	respondJSON(w, 200, map[string]string{"jobId": jobID})

	go processPlaylistAsync(jobID, job, body.URL, isAudio, body.AudioFormat, outputExt, body.Quality, body.Container, body.AudioBitrate, body.ResumeFrom)
}

func processPlaylistAsync(jobID string, job *services.AsyncJob, rawURL string, isAudio bool, audioFormat, outputExt, quality, container, audioBitrate string, resumeFrom int) {
	playlistDir := filepath.Join(config.TempDirs["playlist"], jobID)
	os.MkdirAll(playlistDir, 0755)

	ctx, cancel := context.WithCancel(context.Background())
	processInfo := &services.ProcessInfo{
		TempDir:    playlistDir,
		JobType:    "playlist",
		CancelFunc: cancel,
	}
	services.Global.SetProcess(jobID, processInfo)

	jobs := services.Global.GetJobsByType()
	jobsJSON, _ := json.Marshal(jobs)
	log.Printf("[Queue] Async playlist started. Active: %s", jobsJSON)

	defer cancel()

	playlistInfo, err := services.GetPlaylistInfo(ctx, rawURL, true)
	if err != nil {
		playlistError(jobID, job, processInfo, playlistDir, err)
		return
	}

	if playlistInfo.Count > config.MaxPlaylistVideos {
		playlistError(jobID, job, processInfo, playlistDir, fmt.Errorf("Playlist too large. Maximum %d videos allowed. This playlist has %d videos.", config.MaxPlaylistVideos, playlistInfo.Count))
		return
	}

	totalVideos := playlistInfo.Count
	playlistTitle := playlistInfo.Title
	startIdx := resumeFrom - 1
	if startIdx >= len(playlistInfo.Entries) {
		startIdx = len(playlistInfo.Entries) - 1
	}
	if startIdx < 0 {
		startIdx = 0
	}
	isResuming := startIdx > 0

	var msg string
	if isResuming {
		msg = fmt.Sprintf("resuming from video %d/%d", resumeFrom, totalVideos)
	} else {
		msg = fmt.Sprintf("found %d videos", totalVideos)
	}

	job.Lock()
	job.Status = "downloading"
	job.PlaylistTitle = playlistTitle
	job.TotalVideos = totalVideos
	job.Message = msg
	job.Unlock()

	formatStr := fmt.Sprintf("%s %s", quality, container)
	if isAudio {
		formatStr = audioFormat
	}
	p0 := float64(0)
	services.Global.SendProgress(jobID, "playlist-info", msg, &p0, map[string]interface{}{
		"playlistTitle": playlistTitle, "totalVideos": totalVideos,
		"currentVideo": startIdx, "currentVideoTitle": "", "format": formatStr,
	})

	var downloadedFiles []string
	var failedVideos []services.FailedVideo

	for i := startIdx; i < len(playlistInfo.Entries); i++ {
		if processInfo.IsCancelled() {
			playlistError(jobID, job, processInfo, playlistDir, fmt.Errorf("Download cancelled"))
			return
		}
		if processInfo.IsFinishEarly() {
			log.Printf("[%s] Finishing early after %d videos", jobID, len(downloadedFiles))
			break
		}

		entry := playlistInfo.Entries[i]
		videoNum := i + 1
		videoTitle := orDefault(entry.Title, fmt.Sprintf("Video %d", videoNum))
		videoURL := entry.URL
		if videoURL == "" && entry.ID != "" {
			videoURL = "https://www.youtube.com/watch?v=" + entry.ID
		}
		if videoURL == "" {
			continue
		}

		safeTitle := util.SanitizeFilename(videoTitle)
		if len(safeTitle) > 100 {
			safeTitle = safeTitle[:100]
		}
		videoFile := filepath.Join(playlistDir, fmt.Sprintf("%03d - %s.%s", videoNum, safeTitle, outputExt))

		progress := float64(videoNum-1) / float64(totalVideos) * 100
		job.Lock()
		job.CurrentVideo = videoNum
		job.CurrentVideoTitle = videoTitle
		job.Progress = progress
		job.Message = fmt.Sprintf("downloading %d/%d: %s", videoNum, totalVideos, videoTitle)
		job.Unlock()
		services.Global.SendProgress(jobID, "downloading", fmt.Sprintf("Downloading %d/%d: %s", videoNum, totalVideos, videoTitle),
			&progress, map[string]interface{}{
				"playlistTitle": playlistTitle, "totalVideos": totalVideos,
				"currentVideo": videoNum, "currentVideoTitle": videoTitle,
				"format": formatStr, "failedVideos": failedVideos, "failedCount": len(failedVideos),
			})

		actualURL := videoURL
		isYT := strings.Contains(actualURL, "youtube.com") || strings.Contains(actualURL, "youtu.be")
		var tempPath string

		downloadErr := func() error {
			if isYT {
				result, err := services.DownloadViaYtdlp(ctx, actualURL, fmt.Sprintf("temp_%d", videoNum), services.DownloadOpts{
					IsAudio: isAudio, Quality: quality, Container: container,
					TempDir: playlistDir, ProcessInfo: processInfo, UseProxy: true,
					OnProgress: func(prog float64, speed, eta string) {
						overallProg := (float64(videoNum-1)/float64(totalVideos))*100 + (prog / float64(totalVideos))
						job.Lock()
						job.Progress = overallProg
						job.Speed = speed
						job.ETA = eta
						job.Unlock()
					},
				})
				if err != nil {
					cobaltResult, cobaltErr := services.DownloadViaCobalt(ctx, actualURL, fmt.Sprintf("%s-v%d", jobID, videoNum), isAudio, nil,
						services.CobaltDownloadOpts{OutputDir: playlistDir, MaxRetries: 3, RetryDelay: 2 * time.Second})
					if cobaltErr != nil {
						return cobaltErr
					}
					tempPath = cobaltResult.FilePath
				} else {
					tempPath = result.Path
				}
			} else {
				result, err := services.DownloadViaYtdlp(ctx, actualURL, fmt.Sprintf("temp_%d", videoNum), services.DownloadOpts{
					IsAudio: isAudio, Quality: quality, Container: container,
					TempDir: playlistDir, ProcessInfo: processInfo,
				})
				if err != nil {
					return err
				}
				tempPath = result.Path
			}
			return nil
		}()

		if downloadErr != nil {
			errMsg := downloadErr.Error()
			if errMsg == "Cancelled" || errMsg == "Download cancelled" {
				playlistError(jobID, job, processInfo, playlistDir, downloadErr)
				return
			}
			failedVideos = append(failedVideos, services.FailedVideo{Num: videoNum, Title: videoTitle, Reason: util.ToUserError(errMsg)})
			job.Lock()
			job.FailedVideos = failedVideos
			job.FailedCount = len(failedVideos)
			job.Unlock()
			continue
		}

		if tempPath != "" {
			if _, err := os.Stat(tempPath); err == nil {
				processed, err := services.ProcessVideo(tempPath, videoFile, services.ProcessVideoOpts{
					IsAudio: isAudio, AudioFormat: audioFormat, AudioBitrate: audioBitrate, Container: container, JobID: jobID,
				})
				if err == nil {
					if processed.Skipped && tempPath != videoFile {
						os.Rename(tempPath, videoFile)
					}
					downloadedFiles = append(downloadedFiles, videoFile)
					job.Lock()
					job.VideosCompleted = len(downloadedFiles)
					job.Unlock()
					log.Printf("[%s] Video %d complete", jobID, videoNum)
				}
			}
		}
	}

	if len(downloadedFiles) == 0 {
		playlistError(jobID, job, processInfo, playlistDir, fmt.Errorf("No videos were successfully downloaded"))
		return
	}

	job.Lock()
	job.Status = "zipping"
	job.Progress = 95
	job.Message = fmt.Sprintf("creating zip with %d videos...", len(downloadedFiles))
	job.Unlock()
	p95 := float64(95)
	services.Global.SendProgress(jobID, "zipping", fmt.Sprintf("Creating zip file with %d videos...", len(downloadedFiles)),
		&p95, map[string]interface{}{"playlistTitle": playlistTitle, "totalVideos": totalVideos, "downloadedCount": len(downloadedFiles)})

	zipPath := filepath.Join(config.TempDirs["playlist"], fmt.Sprintf("%s.zip", jobID))
	safePlaylistName := util.SanitizeFilename(orDefault(playlistTitle, "playlist"))

	if err := createZip(zipPath, downloadedFiles); err != nil {
		playlistError(jobID, job, processInfo, playlistDir, fmt.Errorf("Failed to create zip: %v", err))
		return
	}

	os.RemoveAll(playlistDir)

	stat, err := os.Stat(zipPath)
	if err != nil {
		playlistError(jobID, job, processInfo, "", fmt.Errorf("zip file not found after creation"))
		return
	}
	token := randomToken()
	fileName := safePlaylistName + ".zip"

	services.Global.SetBotDownload(token, &services.BotDownload{
		FilePath:      zipPath,
		FileName:      fileName,
		FileSize:      stat.Size(),
		MimeType:      "application/zip",
		CreatedAt:     time.Now(),
		IsWebPlaylist: true,
	})

	job.Lock()
	job.Status = "complete"
	job.Progress = 100
	job.Message = fmt.Sprintf("%d videos ready to download", len(downloadedFiles))
	job.DownloadToken = token
	job.FileName = fileName
	job.FileSize = stat.Size()
	job.FailedVideos = failedVideos
	job.FailedCount = len(failedVideos)
	job.Unlock()

	p100 := float64(100)
	services.Global.SendProgress(jobID, "complete", fmt.Sprintf("%d videos ready!", len(downloadedFiles)),
		&p100, map[string]interface{}{
			"playlistTitle": playlistTitle, "totalVideos": totalVideos,
			"downloadedCount": len(downloadedFiles), "failedVideos": failedVideos,
			"failedCount": len(failedVideos), "downloadToken": token,
		})

	services.Global.ReleaseJob(jobID)
	log.Println("[Queue] Async playlist complete.")
}

func playlistError(jobID string, job *services.AsyncJob, processInfo *services.ProcessInfo, playlistDir string, err error) {
	log.Printf("[%s] Async playlist error: %s", jobID, err)
	alerts.PlaylistFailed(jobID, "", err)
	job.Lock()
	job.Status = "error"
	job.Message = util.ToUserError(err.Error())
	job.Unlock()
	if !processInfo.IsCancelled() {
		services.Global.SendProgressSimple(jobID, "error", util.ToUserError(err.Error()))
	}
	services.Global.ReleaseJob(jobID)
	log.Println("[Queue] Async playlist error.")
	os.RemoveAll(playlistDir)
}

func handlePlaylistStatus(w http.ResponseWriter, r *http.Request) {
	jobID := chi.URLParam(r, "jobId")
	job := services.Global.GetAsyncJob(jobID)
	if job == nil {
		respondJSON(w, 404, map[string]string{"error": "Job not found"})
		return
	}
	respondJSON(w, 200, job.GetPlaylistStatus())
}

func handlePlaylistDownload(w http.ResponseWriter, r *http.Request) {
	token := chi.URLParam(r, "token")
	data := services.Global.GetBotDownload(token)
	if data == nil {
		respondJSON(w, 404, map[string]string{"error": "Download not found or expired"})
		return
	}
	stat, err := os.Stat(data.FilePath)
	if err != nil {
		services.Global.DeleteBotDownload(token)
		respondJSON(w, 404, map[string]string{"error": "File no longer available"})
		return
	}
	asciiFilename := toASCIIFilename(data.FileName)

	w.Header().Set("Content-Type", data.MimeType)
	w.Header().Set("Content-Length", strconv.FormatInt(stat.Size(), 10))
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"; filename*=UTF-8''%s`, asciiFilename, url.PathEscape(data.FileName)))

	f, err := os.Open(data.FilePath)
	if err != nil {
		respondJSON(w, 500, map[string]string{"error": "Failed to open file"})
		return
	}
	defer f.Close()
	io.Copy(w, f)

	data.Downloaded = true
	go func() {
		time.Sleep(30 * time.Second)
		services.Global.DeleteBotDownload(token)
		os.Remove(data.FilePath)
		short := token
		if len(short) > 8 {
			short = short[:8]
		}
		log.Printf("[Playlist] Token %s... cleaned up after download", short)
	}()
}

func createZip(zipPath string, files []string) error {
	f, err := os.Create(zipPath)
	if err != nil {
		return err
	}
	defer f.Close()

	zw := zip.NewWriter(f)
	defer zw.Close()

	for _, filePath := range files {
		entry, err := zw.Create(filepath.Base(filePath))
		if err != nil {
			return err
		}
		src, err := os.Open(filePath)
		if err != nil {
			return err
		}
		io.Copy(entry, src)
		src.Close()
	}
	return nil
}

func ptrFloat(f float64) *float64 {
	return &f
}
